packages/framework/collabspaces/src/collabSpaces.ts

```diff
@@ -439,25 +467,77 @@ export class TempCollabSpaceRuntime
 		return this.matrixInternal;
 	}
 
+	private insertCellIntoReverseMap(
+		map: Map<string, number>,
+		index: number,
+		key: string,
+		value: number,
+	) {
+		// Convert the map to an array of key-value pairs
+		const entries = Array.from(map.entries());
+
+		// Insert the new key-value pair at the specified index
+		entries.splice(index, 0, [key, value]);
+
+		// Clear the original map
+		map.clear();
+
+		// Re-populate the map with the updated items from the array
+		entries.forEach(([localKey], newIndex) => {
+			map.set(localKey, newIndex + 1);
+		});
+	}
+
+	private addCellToReverseMap(row: number, col: number, info: ICellInfo) {
+		// Col and Row == -1 means cell that uniquely identify the rows and columns from the matrix.
+		if (col === -1) {
+			if (this.rowMap.size >= row + 1) {
+				// Cell added
+				this.insertCellIntoReverseMap(
+					this.rowMap,
+					row,
+					info.value as unknown as string,
+					row + 1,
+				);
+			} else {
+				this.rowMap.set(info.value as unknown as string, row + 1);
+			}
+		}
+		if (row === -1) {
+			this.colMap.set(info.value as unknown as string, col + 1);
```

Comment:
I'd expect code to be symmetrical for rows & cols. I'm not sure why col === - 1 case has more sub-cases compared to row === -1 path.

Reply:
the problem once again is the insert - we get one by one - I'll add a comment to make it clear.